hbase-server/src/test/java/org/apache/hadoop/hbase/master/procedure/TestSnapshotProcedureWithLockTimeout.java

```diff
@@ -0,0 +1,149 @@
+    HBaseClassTestRule.forClass(TestSnapshotProcedureWithLockTimeout.class);
+
+  private static HBaseTestingUtil TEST_UTIL;
+  private HMaster master;
+  private TableName TABLE_NAME;
+  private byte[] CF;
+  private String SNAPSHOT_NAME;
+
+  @Before
+  public void setup() throws Exception {
+    TEST_UTIL = new HBaseTestingUtil();
+    Configuration config = TEST_UTIL.getConfiguration();
+    config.setInt("hbase.snapshot.remote.verify.threshold", 1);
+    config.setLong(TakeSnapshotHandler.HBASE_SNAPSHOT_MASTER_LOCK_ACQUIRE_TIMEOUT, 1L);
+    TEST_UTIL.startMiniCluster(3);
+    master = TEST_UTIL.getHBaseCluster().getMaster();
+    TABLE_NAME = TableName.valueOf(Bytes.toBytes("TestSnapshotProcedureWithLockTimeout"));
+    CF = Bytes.toBytes("cf");
+    SNAPSHOT_NAME = "SnapshotProcLockTimeout";
+    final byte[][] splitKeys = new RegionSplitter.HexStringSplit().split(10);
+    Table table = TEST_UTIL.createTable(TABLE_NAME, CF, splitKeys);
+    TEST_UTIL.loadTable(table, CF, false);
+  }
+
+  @After
+  public void teardown() throws Exception {
+    if (this.master != null) {
+      ProcedureTestingUtility.setKillAndToggleBeforeStoreUpdate(master.getMasterProcedureExecutor(),
+        false);
+    }
+    TEST_UTIL.shutdownMiniCluster();
+  }
+
+  @Test
+  public void testTakeZkCoordinatedSnapshot() throws Exception {
+    for (int i = 0; i < 10; i++) {
+      try {
+        // Verify that snapshot creation is not possible because lock could not be
+        // acquired on time. This can be flaky behavior because even though we provide 1ms
+        // as lock timeout, it could still be fast enough and eventually lead to successful
+        // snapshot creation. If that happens, retry again.
+        testTakeZkCoordinatedSnapshot(i);
+        break;
+      } catch (AssertionError e) {
+        LOG.error("Error because of faster lock acquisition. retrying....", e);
+      }
+      assertNotEquals("Retries exhausted", 9, i);
+    }
+  }
+
+  private void testTakeZkCoordinatedSnapshot(int i) throws Exception {
+    SnapshotDescription snapshotOnSameTable =
+      new SnapshotDescription(SNAPSHOT_NAME + i, TABLE_NAME, SnapshotType.SKIPFLUSH);
+    SnapshotProtos.SnapshotDescription snapshotOnSameTableProto =
+      ProtobufUtil.createHBaseProtosSnapshotDesc(snapshotOnSameTable);
+    Thread second = new Thread("zk-snapshot") {
+      @Override
+      public void run() {
+        try {
+          master.getSnapshotManager().takeSnapshot(snapshotOnSameTableProto);
+        } catch (IOException e) {
+          LOG.error("zk snapshot failed", e);
+          fail("zk snapshot failed");
+        }
+      }
+    };
+    second.start();
+
+    Thread.sleep(5000);
+    try {
+      SnapshotTestingUtils.confirmSnapshotValid(TEST_UTIL, snapshotOnSameTableProto, TABLE_NAME,
+        CF);
+      throw new AssertionError("should not have created snapshot " + snapshotOnSameTableProto);
+    } catch (AssertionError e) {
```

Comment:
I suggest we do some refactoring here, catching AssertionError and asserting again looks weird...

And since we use EnvironmentEdgeManager to calculate the deadline, I think you can inject your own EnvironmentEdge for this test and make the behavior stable?

Reply:
Updated to make it clear: https://github.com/apache/hbase/pull/5603/commits/f758519fa90306a434b5b5000ba03676f10cd178

Reply:
I think what we want here is let this.tableLock.tryAcquire throw timeout exception?

You can see the code in EnvironmentEdgeManagerTestHelper, by creating an Exception, you get the full stacktrace so you can determine in which method you are calling the EnvironmentEdge.currentTime, so here I think we can just hack the calls from LockManager.tryAcquire, to make sure that we will not enter the while loop and fail immediately.

Reply:
It would still not be sufficient because we want tryAcquire to return false for our test. It returns false only if main thread exits faster before LockProcedure could be completed. So regardless of our hack of EnvironmentEdgeManager, main thread (i.e. execution of tryAcquire) can still complete (even without entering while loop) after LockProcedure completes execution. If that happens, this if condition would not be true and hence, our test would fail:

```
      if (!proc.isLocked()) {
        LOG.info("Timed out waiting to acquire procedure lock: " + proc.toString());
        release();
        return false;
      }
```

With current retries in place for the test, i have run the test so many times in the loop and i haven't seen it fail anytime so i don't think this would be flaky.

Reply:
OK I got your point, but I think this could make it even easier to make the lock failure, as you can submit a LockProcedure first, but taking snapshot...

Anyway, this has been blocked here for a long time, let's get this in and improve later...